lib/crypto-tools/keyring.js

```diff
@@ -273,4 +281,95 @@ KeyRing.prototype._decrypt = function(data) {
   return JSON.parse(dec) && dec;
 };
 
+/**
+ * Retrieve the bucketKey
+ * @param {String} bucketId - bucket id
+ * @returns {String|null}
+ */
+KeyRing.prototype.generateBucketKey = function(bucketId) {
+  if(!this._mnemonic || !bucketId){
+    return null;
+  }
+  return DeterministicKeyIv.getDeterministicKey(this._mnemonic, bucketId);
+};
+
+/**
+ * Retrieve the fileKey
+ * @param {String} bucketId - bucket id
+ * @param {String} fileId - file id
+ * @returns {DataCipherKeyIv}
+ */
+KeyRing.prototype.generateFileKey = function(bucketId, fileId) {
+  var self = this;
+
+  var bucketKey = self.generateBucketKey(bucketId);
+  if(!bucketKey){
+    return DataCipherKeyIv();
+  }
+
+  var fileKey = DeterministicKeyIv.getDeterministicKey(bucketKey, fileId);
+  return new DeterministicKeyIv(fileKey, fileId);
+};
+
+/**
+ * Read an HD Key from disk
+ * @private
+ */
+KeyRing.prototype._readHDKey = function() {
+  if (!utils.existsSync(this._hdKeyFile)) {
+    return;
+  }
+  var enc = fs.readFileSync(this._hdKeyFile).toString();
+  this._mnemonic = JSON.parse(this._decrypt(enc)).mnemonic;
+};
+
+/**
+ * Generate and save HD Key to disk
+ */
+KeyRing.prototype.generateHDKey = function() {
+
+  if (utils.existsSync(this._hdKeyFile)) {
+    throw new Error('HD Key already exists');
+  }
+
+  this._mnemonic = new Mnemonic(Mnemonic.Words.ENGLISH).toString();
+  this._saveKeyToDisk('.hd_key', { mnemonic: this._mnemonic });
+};
+
+/**
+ * Import mnemonic into KeyRing
+ * @param {String} mnemonic - Mnemonic to transfer
+ */
+KeyRing.prototype.importMnemonic = function(mnemonic) {
```

Comment:
Nomenclature changes + you can use assertions instead of throwing in if blocks to reduce branches